### ISSUE TITLE
Wrong Type Error on missing property when using optional chaining

### ISSUE THREAD

**Issue by Ninerian** [microsoft/TypeScript#35543]
**TypeScript Version:**  3.7.2

**Search Terms:**
- optional chainging
- missing property

**Expected behavior:**
Typescript should handle the optional method as same as optional properties

**Actual behavior:**
Typescript throws a warning: 
```
Property 'foo' does not exist on type 'A | B'.
  Property 'foo' does not exist on type 'B'.(2339)
```

The code itself is doing what I expected. When the method is not found, it uses the right hand value.

**Related Issues:**
No issue found

**Code**
```ts
class A {
    public foo() {
        return 'foo'
    }
    public bar() {
        return 'bar'
    }
}

class B {
    public bar() {
        return 'bar'
    }
}

const x = (F: typeof A | typeof B) => {
    const f = new F();

    return f.foo?.() ?? 'baz';
}

console.log(x(A)); // foo
console.log(x(B)); // baz
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
class A {
    foo() {
        return 'foo';
    }
    bar() {
        return 'bar';
    }
}
class B {
    bar() {
        return 'bar';
    }
}
const x = (F) => {
    var _a, _b, _c;
    const f = new F();
    return _c = (_b = (_a = f).foo) === null || _b === void 0 ? void 0 : _b.call(_a), (_c !== null && _c !== void 0 ? _c : 'baz');
};
console.log(x(A)); // foo
console.log(x(B)); // baz

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/#code/MYGwhgzhAECC0G8CwAoa7oAcCuAjEAlsNAGYD2ZAFAJSKoYPQBOApgC7ZMB20A5OWV70MAX2Hoc+ItFxgmNOmkbpWHbn1lMhS9GJR7UoSDABCihpMLFNC5DsarOPXpu0M9BlMDJcIbaAAe0AC80JQAYgBc0GwAnpgsZCRw0AA+MfGJySa0wQB85hjevv7JoVwsAO7Q4TQA3KjizOxOpAB0AgD8bQqdnRpgAF68DfqNXj4QZCAsbSBkAOaUAZSw1NR10AD0W6QUhpPTs-NLKzkb27uyg0A)

**Comment by dragomirtitian** [comment]
This is not related to optional chaining. Unions have never allowed access to non-common members of union constituent: 

```ts
const x = (o: { a: number } | { b: string }) => {
    o.a // error 
    if (o.a) { // error even though could conceivable be a type guard
        
    }
    if ('a' in o) { // ok type guard
       o.a // o is now narrowed to  { a: number }
    }
}
```
[Playground Link](https://www.typescriptlang.org/play/?ssl=9&ssc=2&pln=1&pc=1#code/MYewdgzgLgBAHjAvDAFCAXDA3jAhpsAVwFsAjAUwCcYBfGAH2xlM2koEswBzWgSiQB82ALAAoGBJggAdLhgB6eTCqUQ1MZJjsAZqhm5+ORcsqrq5AG7kwMKAAsQhLnZihCAGwAmr8MHLsLXFJ3cmZQuSgATwAHUK5CXEpPDU1JFIkadK1dFAByXFytGxBDBSUQAGtbGLiEpKzJfTKpLQgYMBAAd3bE1U7ybygQCRx8dpIKakzxSWmaIA)

**Comment by dragomirtitian** [comment]
Related to #35542

**Comment by Ninerian** [comment]
@dragomirtitian thanks for the explanation and the example. Can be closed from my side.

**Comment by jcalz** [comment]
Duplicate of (the declined 😞) #33736
